js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx

```diff
@@ -59,206 +67,277 @@ export const WorkspaceContext = React.createContext<WorkspaceState>(
+      return await fetch();
+    }, [fetch]),
+    intervalMs: 5000,
+    leading: true,
+  });
 
-  fragment WorkspaceDisplayMetadata on RepositoryMetadata {
-    key
-    value
-  }
+  const {data, loading: loadingCodeLocationStatus} = codeLocationStatusQueryResult;
 
-  fragment WorkspaceLocation on RepositoryLocation {
-    id
-    isReloadSupported
-    serverId
-    name
-    dagsterLibraryVersions {
-      name
-      version
-    }
-    repositories {
-      id
-      ...WorkspaceRepository
-    }
-  }
+  const locations = useMemo(() => getLocations(data), [data]);
+  const prevLocations = useRef<typeof locations>({});
 
-  fragment WorkspaceRepository on Repository {
-    id
-    name
-    pipelines {
-      id
-      name
-      isJob
-      isAssetJob
-      pipelineSnapshotId
-    }
-    schedules {
-      id
-      ...WorkspaceSchedule
-    }
-    sensors {
-      id
-      ...WorkspaceSensor
-    }
-    partitionSets {
-      id
-      mode
-      pipelineName
+  const didInitiateFetchFromCache = useRef(false);
+  const [didLoadCachedData, setDidLoadCachedData] = useState(false);
+
+  const [locationsData, setLocationsData] = React.useState<
+    Record<string, WorkspaceLocationNodeFragment | PythonErrorFragment>
+  >({});
+
+  const getCachedData = useGetCachedData();
+  const getData = useGetData();
+
+  useLayoutEffect(() => {
+    // Load data from the cache
+    if (didInitiateFetchFromCache.current) {
+      return;
     }
-    assetGroups {
-      id
-      groupName
+    didInitiateFetchFromCache.current = true;
+    new Promise(async (res) => {
+      /**
+       * 1. Load the cached code location status query
+       * 2. Load the cached data for those locations
+       * 3. Set the cached data to `locationsData` state
+       * 4. Set prevLocations equal to these cached locations so that we can check if they
+       *  have changed after the next call to codeLocationStatusQuery
+       * 5. set didLoadCachedData to true to unblock the `locationsToFetch` memo so that it can compare
+       *  the latest codeLocationStatusQuery result to what was in the cache.
+       */
+      const data = await getCachedData<CodeLocationStatusQuery>({
+        key: `${localCacheIdPrefix}${CODE_LOCATION_STATUS_QUERY_KEY}`,
+        version: CODE_LOCATION_STATUS_QUERY_VERSION,
+      });
+      const cachedLocations = getLocations(data);
+      const prevCachedLocations: typeof locations = {};
+
+      await Promise.all([
+        ...Object.values(cachedLocations).map(async (location) => {
+          const locationData = await getCachedData<LocationWorkspaceQuery>({
+            key: `${localCacheIdPrefix}${locationWorkspaceKey(location.name)}`,
+            version: LOCATION_WORKSPACE_QUERY_VERSION,
+          });
+          const entry = locationData?.workspaceLocationEntryOrError;
+
+          if (!entry) {
+            return;
+          }
+          setLocationsData((locationsData) =>
+            Object.assign({}, locationsData, {
```

Comment:
I know these reads are async, but they will essentially be immediate, so it might be nice to collect them all (via the Promise.all) and then call setLocationsData once.  If I had ~50 locations, i expect this would rapid-fire setLocationsData 50x and it could lead to some unnecessary thrashing?

Reply:
Yeah I did this in the following PR!